### ISSUE TITLE
False positive when assigning bounded generic to union including the bound

### ISSUE THREAD

**Issue by randolf-scholz** [python/mypy#20516]
The following example is type-safe, but `mypy` reports 6 false positives:

- [pyright playground](https://pyright-play.net/?strict=true&code=CYUwZgBArgDgxgQwM4BcAUCBOBzAXBASwDsUBKCAWgD5CTcAoCJiTEFKTIiLbe%2B0SChCoA2gA18xFAF00BIQFsk%2BADYFRY6eWoQ1G6Q2YQA%2BglXqUIqdIgBeXRbmKk5I24gBiCIFByQPB-jZmMAI3NRazsIEQAqeRAlG3dErz8ApmM4UMtw%2BxEADwgwAHtMCHziQmcEpmT-VJMARkyRVEwbAB8HMJIbez10WKVXJIhAGXJ-I2MAJibwjr7mlFaIvqc4l0Tq0fHAgGYmlvbOrO6I6IGkKo3PLbrjABYZk7mLBaWcmMqr5i8x24BWfaLQ7zbKRfJFEplLjnBI-bZpABsjxkEGeogOp3BxVKtAqa1hNwmAHZAUs0ccUTlYIhUGhcuQITjyjDNikJgAOZHAl4YqnwZDoekFbFQvHxVnwlhsDjQ5x8IA):  no errors
- [pyrefly sandbox](https://pyrefly.org/sandbox/?project=N4IgZglgNgpgziAXKOBDAdgEwEYHsAeAdAA4CeSImMYABAK7EDGqcALgBSoBOA5ojRHSsAlDQC0APgFDEAHXQ1FNLjFZ0uC7j3nyqtVvFYBtABr9BrALrsIBgLZx%2BUCG1OXRkms9cnLchUoA%2BqhOLsYWljQAvF5hNvZwokrJSgDENICg5IDwf-LJgdihrhHRNEYAVLYwDpEptYrp2blBjIXhQpExRvg0YLhcNN2CAgk19Zk56E2KgQCMrUZsXJEAPrFF7SXeHJUOSXXpgDLkE3kATPPFq1sLrEubcTuJdWNHUzSBAMzziytrbVYl5QeoyeNEOxyCABZzhtLmFrrdOhURiC0jQXgFpgBWL43H5XYqdbq9fqDBRA57g6YANmh-1hrm%2BAKJfQG0mGVTgNTBr0CAHYcbd6X8OqUGMw2Ox8KJiayhuTQeMeQAOWl4uGMzpilgcKU9Fmk9nVMaNDHKVTqMkJHSTdAgAA0IDorGgcBI5EQIHSAFVnc5WKQenR0Ixnbh0HBrXo9Vw7KhWIF0HQ7NgYFxJeYhB4pIt-MkVGoND1ZCAAHJJlNcfjAfAAX2L8ntIDIKjAUFIhFYuDsUAo6QACqQW22aGgsHhuoww5AeOo4xAw4R5OkAMowGA0AAWrFYxEcAHo983qG3CH0eHuYOg95hcIw4HvJ%2Bhp7PQ1fozRUAA3VDQVDYWA0I%2Bz5cHOYY0LgxCvq68hkKwG5hmIn6pnA84KDExbvIQMwnPWto1g6qAhhASEAGLQDAFCjjgBBuiANZAA): no errors
- [mypy playground](https://mypy-play.net/?mypy=latest&python=3.12&gist=0a3ad60020dfa795525e6c70230a9e3c): 6 false positives

```python
def upcast(arg: int) -> int:
    return arg

def test[X: int](items: list[X]) -> list[X]:
    _a: list[int] = list(items)           # ✅️
    _b: list[int] = [*items]              # ✅️
    _c: list[int] = [x for x in items]    # ✅️

    _1: list[str] | list[int] = list(items)         # ❌️
    _2: list[int] | list[str] = list(items)         # ❌️
    _3: list[str] | list[int] = [*items]            # ❌️
    _4: list[int] | list[str] = [*items]            # ❌️
    _5: list[str] | list[int] = [x for x in items]  # ❌️
    _6: list[int] | list[str] = [x for x in items]  # ❌️
    _7: list[str] | list[int] = [upcast(x) for x in items]  # ✅️
    _8: list[int] | list[str] = [upcast(x) for x in items]  # ✅️
    return items
```

context: https://github.com/python/mypy/pull/20419#discussion_r2658824097
